plutus-contract/test/Spec/Contract/Tx/Constraints/MustUseOutputAsCollateral.hs

```diff
@@ -0,0 +1,326 @@
+    in checkPredicateOptions defaultCheckOptions
+      ("Successful use of offchain and onchain mustUseOutputAsCollateral for a single " ++
+       "collateral input")
+      (assertValidatedTransactionCount 1 .&&.
+        (assertUnbalancedTx contract
+        (Trace.walletInstanceTag w1)
+        (\unT -> length (Tx.getCardanoTxCollateralInputs $ Tx.EmulatorTx $ unT ^. OffCon.tx) ==  1)
+        "correct number of collateral inputs"))
+      (void $ trace contract)
+
+-- | Valid scenario using offchain and onchain constraint mustUseOutputAsCollateral to select
+-- multiple utxos to use as collateral input
+multipleUseOfMustUseOutputAsCollateral :: SubmitTx -> PSU.Language -> TestTree
+multipleUseOfMustUseOutputAsCollateral submitTxFromConstraints lc =
+    let contract = mustUseOutputAsCollateralContract submitTxFromConstraints
+                  lc maximumCollateralInputs w1PaymentPubKeyHash
+    in checkPredicateOptions defaultCheckOptions
+      ("Successful use of offchain and onchain mustUseOutputAsCollateral for the maximum " ++
+       "number of allowed collateral inputs")
+      (assertValidatedTransactionCount 1 .&&.
+        (assertUnbalancedTx contract
+        (Trace.walletInstanceTag w1)
+        (\unT ->
+          length (Tx.getCardanoTxCollateralInputs $ Tx.EmulatorTx $ unT ^. OffCon.tx)
+            ==  fromIntegral maximumCollateralInputs)
+        "correct number of collateral inputs"))
+      (void $ trace contract)
+
+-- | Valid scenario when using offchain and onchain constraint mustUseOutputAsCollateral with other
+-- wallet's utxo
+usingMustUseOutputAsCollateralWithOtherWalletUtxo
+    :: SubmitTx
+    -> PSU.Language
+    -> TestTree
+usingMustUseOutputAsCollateralWithOtherWalletUtxo submitTxFromConstraints lc =
+    let numberOfCollateralInputs = 2
+        contract = mustUseOutputAsCollateralContract submitTxFromConstraints lc
+                    numberOfCollateralInputs w1PaymentPubKeyHash
+        traceWithW2Signing = do
+            Trace.setSigningProcess w1 (Just $ signPrivateKeys
+                [paymentPrivateKey $ walletToMockWallet' w1,
+                paymentPrivateKey $ walletToMockWallet' w2])
+            void $ Trace.activateContractWallet w1 contract
+            Trace.nextSlot
+    in checkPredicateOptions defaultCheckOptions
+    "Successful use of offchain and onchain mustUseOutputAsCollateral when using other wallet's utxo"
+    (assertValidatedTransactionCount 1 .&&.
+    (assertUnbalancedTx contract
+    (Trace.walletInstanceTag w1)
+    (\unT ->
+        length (Tx.getCardanoTxCollateralInputs $ Tx.EmulatorTx $ unT ^. OffCon.tx)
+        ==  fromIntegral numberOfCollateralInputs)
+    "correct number of collateral inputs"))
+    (void traceWithW2Signing)
+
+-- | Ledger validation error scenario when offchain constraint mustUseOutputAsCollateral is used
+-- with a script's utxo
+usingMustUseOutputAsCollateralWithScriptUtxo :: SubmitTx -> PSU.Language -> TestTree
```

Comment:
For consistency, maybe call this `ledgerValidationErrorWhenUsingMustUseOutputAsCollateralWithScriptUtxo`

Reply:
Yup, originally I didn't know what the outcome would be :smile: